Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp

```diff
@@ -5943,19 +5946,33 @@ IGNORE_CLANG_WARNINGS_END
     {
         JSGlobalObject* globalObject = m_graph.globalObjectFor(m_origin.semantic);
         ArrayMode arrayMode = m_node->arrayMode();
+
+        // This function is used by `ArrayAt` but it supports only Int32, Contiguous, Double.
+        ASSERT(m_node->op() != ArrayAt || (arrayMode.type() == Array::Int32 || arrayMode.type() == Array::Contiguous || arrayMode.type() == Array::Double));
+
         switch (arrayMode.type()) {
         case Array::Int32:
         case Array::Contiguous: {
-            LValue index = lowInt32(m_graph.varArgChild(m_node, 1));
+            LValue originalIndex = lowInt32(m_graph.varArgChild(m_node, 1));
             LValue storage = lowStorage(m_graph.varArgChild(m_node, 2));
 
+            LValue arrayLength = m_out.load32NonNegative(storage, m_heaps.Butterfly_publicLength);
+            LValue index = m_node->op() == ArrayAt
+                ? m_out.select(m_out.lessThan(originalIndex, m_out.int32Zero), m_out.add(arrayLength, originalIndex), originalIndex)
+                : originalIndex;
+
             IndexedAbstractHeap& heap = arrayMode.type() == Array::Int32 ?
                 m_heaps.indexedInt32Properties : m_heaps.indexedContiguousProperties;
 
             LValue base = lowCell(m_graph.varArgChild(m_node, 0));
 
             if (arrayMode.isInBounds()) {
-                LValue result = m_out.load64(baseIndexWithProvenValue(heap, storage, index, m_graph.varArgChild(m_node, 1)));
+                if (m_node->op() == ArrayAt)
+                    speculate(OutOfBounds, noValue(), nullptr, m_out.aboveOrEqual(index, arrayLength));
```

Comment:
Doesn't CheckInBounds force you to OSR exit here when the index is negative? If I recall correctly, CheckInBounds always exits on negative indices. I think you need to teach check in bounds about `at`s case or just assume the index is non-negative for the in bounds case. If you do the latter I think you just don't need this check, although it might be worth it as a debug assert.

Reply:
My current patch does not use the `CheckInBounds` node. I didn't know that.

Also, in my opinion, we should not assume that the index is non-negative (https://github.com/WebKit/WebKit/pull/38331#discussion_r1917851552).

Ideally, I think we should introduce a new `CheckInBoundsNegative` node and insert it in `DFGSSALoweringPhase`. This could also be used for the FTL for `String#At` that was introduced in https://commits.webkit.org/287653@main.

However, adding a new node would involve a amount of changes. So, I’d prefer to leave this patch as is, mark it with a `FIXME`( https://bugs.webkit.org/show_bug.cgi?id=286056 ), and handle this in a separate patch.

As far as I understand, this patch should work as expected as is.

Reply:
IIRC, one of the reasons to use `at` is that it works with containers that don't support array access. But upon further searching those don't seem to exist. In that case, it's not with adding an optimization until we know folks use `at` in that way.

Ah I was mistaken, I thought CheckInBounds was automatically inserted for arrayMode nodes in the FTL. Seems like that's not the case. A FIXME is fine. Although I think I'd implement `at` with a `Select` node rather than a separate `CheckInBoundsNegative`. Which, thinking about it more would mean you don't have to have a `ArrayAt` node at all. So maybe that's the right approach for this whole PR...

Reply:
>  Although I think I'd implement at with a Select node rather than a separate CheckInBoundsNegative. Which, thinking about it more would mean you don't have to have a ArrayAt node at all. So maybe that's the right approach for this whole PR...

Thank you for the suggestion! This is my first time hearing about the `Select` node. Could you explain a bit more about how it works and how it could be applied here?